Mathlib/Algebra/GroupPower/Lemmas.lean

```diff
@@ -2,35 +2,1316 @@
+    (multiplesAddHom A).symm f = f 1 :=
+  rfl
+#align multiples_add_hom_symm_apply multiples_add_hom_symm_apply
+
+@[simp]
+theorem zmultiples_add_hom_apply [AddCommGroup A] (x : A) (n : ℤ) :
+    zmultiplesAddHom A x n = n • x :=
+  rfl
+#align zmultiples_add_hom_apply zmultiples_add_hom_apply
+
+@[simp]
+theorem zmultiples_add_hom_symm_apply [AddCommGroup A] (f : ℤ →+ A) :
+    (zmultiplesAddHom A).symm f = f 1 :=
+  rfl
+#align zmultiples_add_hom_symm_apply zmultiples_add_hom_symm_apply
+
+/-!
+### Commutativity (again)
+
+Facts about `semiconj_by` and `commute` that require `zpow` or `zsmul`, or the fact that integer
+multiplication equals semiring multiplication.
+-/
+
+
+namespace SemiconjBy
+
+section
+
+variable [Semiring R] {a x y : R}
+
+@[simp]
+theorem cast_nat_mul_right (h : SemiconjBy a x y) (n : ℕ) : SemiconjBy a ((n : R) * x) (n * y) :=
+  SemiconjBy.mul_right (Nat.commute_cast _ _) h
+#align semiconj_by.cast_nat_mul_right SemiconjBy.cast_nat_mul_right
+
+@[simp]
+theorem cast_nat_mul_left (h : SemiconjBy a x y) (n : ℕ) : SemiconjBy ((n : R) * a) x y :=
+  SemiconjBy.mul_left (Nat.cast_commute _ _) h
+#align semiconj_by.cast_nat_mul_left SemiconjBy.cast_nat_mul_left
+
+@[simp]
+theorem cast_nat_mul_cast_nat_mul (h : SemiconjBy a x y) (m n : ℕ) :
+    SemiconjBy ((m : R) * a) (n * x) (n * y) :=
+  (h.cast_nat_mul_left m).cast_nat_mul_right n
+#align semiconj_by.cast_nat_mul_cast_nat_mul SemiconjBy.cast_nat_mul_cast_nat_mul
+
+end
+
+variable [Monoid M] [Group G] [Ring R]
+
+@[simp, to_additive]
+theorem units_zpow_right {a : M} {x y : Mˣ} (h : SemiconjBy a x y) :
+    ∀ m : ℤ, SemiconjBy a ↑(x ^ m) ↑(y ^ m)
+  | (n : ℕ) => by simp only [zpow_ofNat, Units.val_pow_eq_pow_val, h, pow_right]
+  | -[n+1] => by simp only [zpow_negSucc, Units.val_pow_eq_pow_val, units_inv_right, h, pow_right]
+#align semiconj_by.units_zpow_right SemiconjBy.units_zpow_right
+#align add_semiconj_by.units_zsmul_right AddSemiconjBy.addUnits_zsmul_right
+
+variable {a b x y x' y' : R}
+
+@[simp]
+theorem cast_int_mul_right (h : SemiconjBy a x y) (m : ℤ) : SemiconjBy a ((m : ℤ) * x) (m * y) :=
+  SemiconjBy.mul_right (Int.commute_cast _ _) h
+#align semiconj_by.cast_int_mul_right SemiconjBy.cast_int_mul_right
+
+@[simp]
+theorem cast_int_mul_left (h : SemiconjBy a x y) (m : ℤ) : SemiconjBy ((m : R) * a) x y :=
+  SemiconjBy.mul_left (Int.cast_commute _ _) h
+#align semiconj_by.cast_int_mul_left SemiconjBy.cast_int_mul_left
+
+@[simp]
+theorem cast_int_mul_cast_int_mul (h : SemiconjBy a x y) (m n : ℤ) :
+    SemiconjBy ((m : R) * a) (n * x) (n * y) :=
+  (h.cast_int_mul_left m).cast_int_mul_right n
+#align semiconj_by.cast_int_mul_cast_int_mul SemiconjBy.cast_int_mul_cast_int_mul
+
+end SemiconjBy
+
+namespace Commute
+
+section
+
+variable [Semiring R] {a b : R}
+
+@[simp]
+theorem cast_nat_mul_right (h : _root_.Commute a b) (n : ℕ) : _root_.Commute a ((n : R) * b) :=
+  SemiconjBy.cast_nat_mul_right h n
```

Comment:
That's really annoying that we can't use dot notation here. Lean incorrectly interprets `h.cast_nat_mul_right` as `Commute.cast_nat_mul_right` and complains that our definition is recursive. Reported here https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/incorrect.20assumption.20about.20dot.20notation.20declaration.20name/near/316626502